service/core/kafka/server/src/main/scala/com/lightbend/lagom/internal/broker/kafka/TopicProducerActor.scala

```diff
@@ -52,79 +88,108 @@ private[lagom] class TaggedOffsetProducerActor[Message](
+      val (killSwitch, streamDone) = backoffSource
+        .viaMat(KillSwitches.single)(Keep.right)
+        .toMat(Sink.ignore)(Keep.both)
+        .run()
 
-    case (None, _) =>
-      // log is impacted by kafkaConfig.serviceName
-      kafkaConfig.serviceName match {
-        case Some(serviceName) => log.error("Unable to locate Kafka service named [{}]", serviceName)
-        case None              => log.error("Unable to locate Kafka brokers URIs")
-      }
+      shutdown = Some(killSwitch)
+      streamDone.pipeTo(self)
+    }
 
-      context.stop(self)
-  }
+    case EnsureActive(_) =>
+    // Yes, we are active
 
-  private def active: Receive = generalHandler.orElse {
     case Done =>
+      // This `Done` is materialization of the `Sink.ignore` above.
       log.info("Kafka producer stream for topic {} was completed.", topicId)
       context.stop(self)
+
+    case Status.Failure(e) =>
+      throw e
   }
 
-  private def run(tag: String, endpoints: String, dao: OffsetDao) = {
-    val readSideSource = eventStreamFactory(tag, dao.loadedOffset)
+  /**
+   * Every time we want to re/start the stream we need to locate the brokers and load the latest offset.
+   * The returned future can fail when the offset store is unavailable or times out or when the brokers list
+   * is empty or unconfigured, etc... In either case, the stream can't be built
+   */
+  private def eventualBrokersAndOffset(tagName: String): Future[(String, OffsetDao)] = {
+    // TODO: review the OffsetStore API. `prepare()` does a lot more than we want here. Ideally, prepare would be
+    // used when this actor starts and here we would only query for the latest offset.
+    val daoFuture: Future[OffsetDao] = offsetStore.prepare(s"topicProducer-$topicId", tagName)
+
+    // null or empty strings become None, otherwise Some[String]
+    def strToOpt(str: String) =
+      Option(str).filter(_.trim.nonEmpty)
+
+    // can be `None` if: not found using service locator (given serviceName)
+    // and no fallback `brokers` are setup in `config`
+    val serviceLookupFuture: Future[Option[String]] =
+      kafkaConfig.serviceName match {
+        case Some(name) =>
+          locateService(name)
+            .map { uris =>
+              strToOpt(UriUtils.hostAndPorts(uris))
+            }
+
+        case None =>
+          Future.successful(strToOpt(kafkaConfig.brokers))
+      }
+
+    val brokerList: Future[String] = serviceLookupFuture.flatMap {
+      case Some(brokers) => Future.successful(brokers)
+      case None =>
+        kafkaConfig.serviceName match {
+          case Some(serviceName) =>
+            val msg = s"Unable to locate Kafka service named [$serviceName]. Retrying..."
+            log.error(msg)
+            Future.failed(new IllegalArgumentException(msg))
+          case None =>
+            val msg = "Unable to locate Kafka brokers URIs. Retrying..."
+            log.error(msg)
+            Future.failed(throw new RuntimeException(msg))
```

Comment:
throw.

Reply:
```suggestion
            Future.failed(new RuntimeException(msg))
```

Reply:
Yeah, shuffling code around I forgot this. Good catch!

Reply:
oh, @dwijnand I did the change locally already. thanks, though.